src/Domain/WeightedElementDistribution.cpp

```diff
@@ -0,0 +1,240 @@
+            // if we were to add one more element, so we add the current element
+            // to the current proc
+            cost_spent_on_proc += element_cost;
+            cost_remaining -= element_cost;
+            num_elements_distributed_to_proc++;
+            total_elements_distributed_to_proc++;
+            current_element_of_current_block++;
+          }
+        }
+      }
+
+      // add a proc and its element allowance for the current block
+      block_element_distribution_.at(current_block)
+          .emplace_back(std::make_pair(global_proc_number,
+                                       num_elements_distributed_to_proc));
+      if (current_element_of_current_block >= num_elements_current_block) {
+        // if we're done assigning elements from the current block, move on to
+        // the elements in the next block
+        ++current_block;
+        current_element_of_current_block = 0;
+      }
+    }
+  }
+}
+
+template <size_t Dim>
+std::vector<std::vector<double>>
+WeightedBlockZCurveProcDistribution<Dim>::get_cost_by_element_by_block(
+    const std::vector<Block<Dim>>& blocks,
+    const std::vector<std::array<size_t, Dim>>& initial_refinement_levels,
+    const std::vector<std::array<size_t, Dim>>& initial_extents,
+    const Spectral::Quadrature quadrature) {
+  // elemental costs correspond to elements in Z-curve order
+  std::vector<std::vector<double>> cost_by_element_by_block(blocks.size());
+
+  for (size_t block_number = 0; block_number < blocks.size(); block_number++) {
+    const auto& block = blocks[block_number];
+    const auto initial_ref_levs = initial_refinement_levels[block.id()];
+    const std::vector<ElementId<Dim>> element_ids =
+        initial_element_ids_in_z_curve_order(block.id(), initial_ref_levs);
+    const size_t grid_points_per_element = alg::accumulate(
+        initial_extents[block.id()], 1_st, std::multiplies<size_t>());
+
+    cost_by_element_by_block[block_number].reserve(element_ids.size());
+
+    // compute the minimum grid spacing (in Frame::Grid) and cost of each
+    // element
+    for (const auto& element_id : element_ids) {
+      Mesh<Dim> mesh = ::domain::Initialization::create_initial_mesh(
+          initial_extents, element_id, quadrature);
+      Element<Dim> element = ::domain::Initialization::create_initial_element(
+          element_id, block, initial_refinement_levels);
+      ElementMap<Dim, Frame::Grid> element_map{
+          element_id, block.is_time_dependent()
+                          ? block.moving_mesh_logical_to_grid_map().get_clone()
+                          : block.stationary_map().get_to_grid_frame()};
+
+      tnsr::I<DataVector, Dim, Frame::ElementLogical> logical_coords{};
+      domain::Tags::LogicalCoordinates<Dim>::function(
+          make_not_null(&logical_coords), mesh);
+
+      tnsr::I<DataVector, Dim, Frame::Grid> grid_coords{};
+      domain::Tags::MappedCoordinates<
```

Comment:
If you look at the function for this tag, all it does is
```cpp
*target_coords = element_map(source_coords);
```
I think you can just do that yourself rather than calling into the compute tag function. It will look simpler.

Reply:
It's funny you say this because I also almost just stuck in this line instead of calling the function and the reason I called the function (regardless of how simple) is just for maintainability like if for whatever reason the function body changed, then that change wouldn't also be reflected in the use in `DgElementArray.hpp`. I don't expect a wild change, it just feels more maintainable to reuse a piece that *could* be updated. But, I also don't have a strong opinion on this, just wanted to explain why there could be a benefit to calling the function, instead

Reply:
I totally get your reasoning, but this is a lot of templates and tags for something that doesn't actually use tags, so I think just using `element_map(source_coords)` is clear enough (same as `logical_coords` above)

Reply:
yes please, use the free functions :)